azure/services/aso/aso.go

```diff
@@ -122,20 +122,18 @@ func (r *reconciler[T]) CreateOrUpdateResource(ctx context.Context, spec azure.A
 				// update instead of returning early.
 				adopt = true
 			case cond.Reason == conditions.ReasonReconciling.Name:
-				readyErr = azure.NewOperationNotDoneError(&infrav1.Future{
+				err := azure.NewOperationNotDoneError(&infrav1.Future{
 					Type:          createOrUpdateFutureType,
 					ResourceGroup: existing.GetNamespace(),
 					Name:          existing.GetName(),
 				})
+				return zero, azure.WithTransientError(err, requeueInterval)
 			default:
-				readyErr = fmt.Errorf("resource is not Ready: %s", conds[i].Message)
-			}
-
-			if readyErr != nil {
+				err := fmt.Errorf("resource is not Ready: %s", conds[i].Message)
```

Comment:
The changes around here are because I noticed that before, if an ASO resource got stuck in a failed state, attempting to fix it by updating the CAPZ spec would never work because we were always returning an error here without calling `Parameters` and updating the ASO resource. These changes defer the equivalent check to after `Parameters` is called and generates no diff.

Reply:
Merged in #4148